### ISSUE TITLE
sql/parser: replace vitess grammar with postgres grammar

### ISSUE THREAD

**Issue by petermattis** [cockroachdb/cockroach#1670]
TL;DR? I think we should replace the vitess SQL grammar with a trimmed down postgres SQL grammar.

While working on SQL expression evaluation I started running into the limitations of the vitess SQL parser: incorrect handling of booleans, forcing boolean expressions where value expressions are allowed, etc. The short summary is that vitess parses a subset of MySQL. Out of the box it didn't handle various DDL statements such as CREATE and ALTER. I've made a number of fixes to the grammar, but I believe this is a game of whack-a-mole and that it is time to switch to a better base.

The [Postgres SQL grammar](https://github.com/postgres/postgres/blob/master/src/backend/parser/gram.y) weighs in at a hefty 14,850 lines. Interestingly, this is fairly close to the size of the [MySQL grammar](http://osxr.org/mysql/source/sql/sql_yacc.yy). Strip out the C code (used for semantic actions) and the Postgres grammar drops down to 5,658 lines. The grammar includes a lot of Postgres extensions (such as the ability to [create new types](http://www.postgresql.org/docs/9.4/static/sql-createtype.html) and [operators](http://www.postgresql.org/docs/9.4/static/sql-createoperator.html)) which can also be trimmed to bring the grammar down to ~3000 lines of code.

Why Postgres vs MySQL? MySQL is certainly the more popular database, but Postgres attempts to follow the SQL standard. It is very unlikely that our brand of SQL will be 100% compatible with either Postgres or MySQL. We certainly don't want to implement all of the crazy Postgres extensions (see custom types and operators) not do we want to limit ourselves to their SQL (e.g. we almost certainly want interleaved indexes). That said, it seems preferable to start with a base that pays heed to the standards.

The Postgres grammar is mostly compatible with [our syntax tree](https://github.com/cockroachdb/cockroach/blob/master/sql/parser/ast.go), though we'll definitely need to extend the structures in certain places. We will also have to replace the tokenizer. But the vast majority of the work will be in writing new semantic actions for the grammar rules. The most unfortunate irritation of switching to the Postgres grammar is that the go yacc tool has a hardcoded limit of [2000 states](https://github.com/golang/go/blob/master/src/cmd/yacc/yacc.go#L64). Even after the trimming mentioned above the grammar contains 2200+ states. It's conceivable that further trimming will get this under 2000 states, but in the meantime the easiest solution is to fork the yacc tool and adjust the hardcoded values.

I spent last week exploring the feasibility of the proposed switch, wrapping my head around the postgres grammar and tokenizer. I'm confident the switch is feasible in a short time frame and will save us lots of headaches down the road.

**Comment by spencerkimball** [comment]
I'm fine with this. Can you give a couple of examples of pointed
differences between the two grammars? Are people going to be stumbling over
them?

**Comment by petermattis** [comment]
Hard to point to concise examples of differences between the two grammars as the Postgres one is much more complete. I think the biggest win is that [SELECT statements](https://github.com/postgres/postgres/blob/master/src/backend/parser/gram.y#L9832) and expression handling in the Postgres grammar strive hard to be standards compliant. See the [a_expr](https://github.com/postgres/postgres/blob/master/src/backend/parser/gram.y#L11429) and [b_expr](https://github.com/postgres/postgres/blob/master/src/backend/parser/gram.y#L11839) rules in the Postgres grammar vs the [value_expression](https://github.com/cockroachdb/cockroach/blob/master/sql/parser/sql.y#L879) and [boolean_expression](https://github.com/cockroachdb/cockroach/blob/master/sql/parser/sql.y#L734) rules in the vitess grammar. Also note the [reserved_keyword](https://github.com/postgres/postgres/blob/master/src/backend/parser/gram.y#L13953) rule in the Postgres grammar, a detail which the vitess grammar mostly ignores.

Most of the differences will not be things people stumble over immediately, but they are certainly present and would make our SQL dialect more distinct than it needs to be.

**Comment by petermattis** [comment]
As of #1727, this is now done.
